keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,172 @@
+
+1) These reconciliation phases are where Kubernetes performs actual work that can be traced
+2) Prevents traces from spanning the entire life of an object, which can be indefninite
+3) Provides a consistent, concrete duration for the root span which remains applicable across object types
+4) The time required for Kubernetes controllers to reconcile desired and actual state is, by definition, a measurement of either latency experienced by end-users in the fulfillment of their requests, or down-time due to a disruption
+
+#### Root spans
+
+In the standard model for distributed tracing, there exists a span in each trace that all other spans are descendents of and which extends the length of the entire trace, called the `root span`. 
+
+The Kubernetes component that kicks off an operation might not be the same component that ends it. In this proposal, when we are at the point where we want to end a root span, we craft a span to export which acts as the root span for the trace. For example, when the kubelet updates a pod from `Pending` to `Running`, it creates a root span using the start time of the pod as the start, and the current time as the end.
+
+
+#### Context propagation
+
+To correlate work done between components as belonging to the same trace, we must pass span context across process boundaries. In traditional distributed systems, this context can be passed down through RPC metadata or HTTP headers. Kubernetes, however, due to its watch-based nature, requires us to attach trace context directly to the target object. 
+
+In this proposal, we choose to propagate this span context as an encoded string an object annotation called `trace.kubernetes.io/context`. This annotation value is regenerated and replaced when an object's trace ends, to achieve the desired behavior from [section one](#trace-lifecycle). 
+
+This proposal chooses to use annotations as a less invasive alternative to adding a field to object metadata, but as this proposal matures, adding trace context to the official API should be considered. 
+
+
+### In-tree changes
+
+#### Vendoring the OpenCensus trace framework
+
+This KEP proposes the use of the [OpenCensus tracing framework](https://opencensus.io/) to create and export spans to configured backends. The OpenCensus framework was chosen for various reasons: 
+
+1) Provides concrete, tested implementations for creating and exporting spans to diverse backends, rather than providing an API specification, as is the case with [OpenTracing](https://opentracing.io/specification/)
+2) [Provides an agent](https://github.com/census-instrumentation/opencensus-service) which enables lazy configuration for exporters, batching of spans, and other features 
```

Comment:
This is likely going to need a sig-architecture discussion, as I'm not sure this heavy of a dependency is something we want to carry long term. I don't know enough about OpenCensus, is this really a required component?

Reply:
The OpenCensus agent is not a required component for this proposal, and more of possible route we could take in the future (will update proposal to reflect this). It's possible to export spans from the components themselves directly to the configured backends (which is what our current implementation work has been doing).

Reply:
I see the reason and benefit of extracting this into the sidecar, but I'm not seeing this feature ever leaving preview or alpha state without this issue being resolved. I'm ok with it at this stage, but I want to have mentioned it upfront, as I have doubts with sig-architecture approving this even as an optional feature, as it's a significant change to how Kubernetes is used/deployed/operated. The OpenCensus team even encountered problems suggesting to deploy the agent as DaemonSets. See "Open Questions" here: https://docs.google.com/document/d/1U2McyGwPIm0win_0uNQqUlPJrrQh1WH5J4m8q8KQyv4/edit#heading=h.rgbw704usq10